client/allocrunner/alloc_runner.go

```diff
@@ -1233,6 +1233,84 @@ func (ar *allocRunner) Signal(taskName, signal string) error {
 	return err.ErrorOrNil()
 }
 
+// Reconnect synchronizes client state between an incoming unknown allocation and
+// the current allocRunner and appends a reconnect event to each task if alloc is running.
+func (ar *allocRunner) Reconnect(update *structs.Allocation) {
+	// Guard against the server sending update requests faster than the state
+	// gets propagated back to them.
+	if ar.hasPendingReconnect() {
+		ar.logger.Trace("skipping due to recent reconnect")
+		return
+	}
+
+	ar.maybeAppendReconnectEvent()
+
+	// Populate the alloc with updated events.
+	clientAlloc := ar.clientAlloc(ar.TaskRunnerTaskStates())
+
+	// Sync the incoming server state with the actual client state including the newly minted event.
+	update.ClientStatus = clientAlloc.ClientStatus
+	update.ClientDescription = clientAlloc.ClientDescription
+	update.TaskStates = clientAlloc.TaskStates
+
+	// Update the local copy of alloc.
+	if err := ar.stateDB.PutAllocation(update); err != nil {
+		ar.logger.Error("error reconnecting alloc when persisting updated alloc locally", "error", err, "alloc_id", update.ID)
+		return
+	}
+
+	// Sync the alloc runner. Important for ensuring the AllocModifyIndex is propagated.
+	// Do not call setAlloc on TaskRunners. That will restart the alloc.
+	ar.setAlloc(update)
+
+	// Update the server.
+	ar.stateUpdater.AllocStateUpdated(update)
+}
+
+// hasPendingReconnect checks the current alloc TaskRunnerTaskStates to see if a recent
+// reconnect event was appended. This is useful for allowing the client state to
+// have time to propagate to the servers before attempting another reconnect.
+func (ar *allocRunner) hasPendingReconnect() bool {
+	now := time.Now()
```

Comment:
Meant to leave a general question here for you @tgross.  Assuming the later comments are addressed, Is this kind of approach ok? Do we have some other type of back-off mechanism that is preferred? The hard-coded interval also feels weird 🤔  but I don't want to add more config.

Reply:
Hm, I may not understand the reasoning here... doesn't this risk dropping updates from the client? We don't typically have backoff on that other than the batching of changes that we already do elsewhere (in `node_update.go`)

Reply:
Yeah it might now that I am getting deeper into the AllocState() code. The reasoning is the server was sending a second reconnect request before the first finished updating and propagating back to the server. Verified by it having a different AllocModifyIndex which I think is right? So I ended up with 2 reconnect events the TaskEvents. I'm trying to stop that.

Reply:
Ok, after re-reviewing after you addressed the other comments...

The thing that jumps out at me here is that we get the same thing (multiple task events) if we have multiple tasks in the task group, right? Should we really be using the `TaskEvent` to trigger alloc-level behavior like this at all?

Reply:
It's not that it is my goal to trigger alloc level behavior this way. I was just trying to get an event in the `Events` for each task. Once I did that, I noticed this has the beneficial side effect of updating the server. 

It does feel right to me that you'd want an event on each task as they are reconnected, but the downside is that each one of those updates will trigger `TaskStateUpdated` which bubbles up multiple times for the same alloc all the way to the client. 

I don't see anything that writes to `Events` outside of the task runner, and I don't see any sort of analog on the `Allocation` struct itself where we could log an Allocation event.

Reply:
> It does feel right to me that you'd want an event on each task as they are reconnected, but the downside is that each one of those updates will trigger `TaskStateUpdated` which bubbles up multiple times for the same alloc all the way to the client.

Yeah we're running into a semantic mismatch for "task events", because we don't have "alloc events" and this is very much an allocation-level event and not a task event. I've got the same problem going on with the CSI hook and being unable to surface mount/unmount to tasks events nicely.

But I think having the alloc-level event show up on multiple tasks is fine. That leaves two different problems to solve:

* Collapsing multiple `Reconnect` events for the allocation into a single _set_ of task events. That's more-or-less solved with this backoff loop but I'm wondering if it could be more straightforwardly solved by an allocrunner-level lock, similar to how we treat all the various alloc runner hooks. But I might be misunderstanding the source of having multiple in-flight reconnects?
* Making sure that the server doesn't treat each of those task events as a new event to kick off another eval. It's not clear to me whether this is an actual problem... do those evals just get collapsed on the eval broker?

Reply:
I think there must be 2 things going on here that I need to suss out. The second reconnect event is reaching the client with a higher `AllocModifyIndex` but with the `ClientStatus` still set to _unknown_. That mean's it's not being triggered by the event emission, or else the `ClientStatus` wouldn't be _unknown_. I didn't connect those dots until this conversation so I've got to go figure that out. 

Regarding collapsing the multiple `Reconnect` events into a single set, my previous implementation using `AppendEvent` and `clientUpdate` was attempting to do just that, and was modeled after the health hook's `SetHealth` implementation. However, as you noticed, it wasn't being very smart about handling the number of iterations over `TaskState` etc. I'll take a stab at getting that correct.

Regarding whether new evals get kicked off...now I'm wondering if that is the genesis of the second reconnect 🤔

Reply:
I like the architecture of the solution you've got in [031a17e](https://github.com/hashicorp/nomad/pull/12133/commits/031a17ed45d8cb5d8b42100a98d41ae61a296c14). Going to mark this thread as resolved.